File: android/src/main/kotlin/com/kurenai7968/volume_controller/VolumeControllerPlugin.kt
package com.kurenai7968.volume_controller

import android.content.Context
import android.media.AudioManager
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

class VolumeControllerPlugin : FlutterPlugin, MethodCallHandler {
  private lateinit var volumeController: VolumeController
  private lateinit var methodChannel: MethodChannel
  private lateinit var eventChannel: EventChannel

  override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    val context = flutterPluginBinding.applicationContext
    val audioManager = context.getSystemService(Context.AUDIO_SERVICE) as AudioManager
    val volumeListener = VolumeListener(context, audioManager)
    volumeController = VolumeController(audioManager)

    // Register EventChannel
    eventChannel = EventChannel(flutterPluginBinding.binaryMessenger, ChannelName.EVENT_CHANNEL)
    eventChannel.setStreamHandler(volumeListener)

    // Register MethodChannel
    methodChannel = MethodChannel(flutterPluginBinding.binaryMessenger, ChannelName.METHOD_CHANNEL)
    methodChannel.setMethodCallHandler(this)
  }

  override fun onMethodCall(call: MethodCall, result: Result) {
    when (call.method) {
      MethodName.GET_VOLUME -> {
        val volume = volumeController.getVolume()
        result.success(volume)
      }
      MethodName.SET_VOLUME -> {
        val volume: Double = call.argument(MethodArgument.VOLUME)!!
        val showSystemUI: Boolean = call.argument(MethodArgument.SHOW_SYSTEM_UI)!!

        volumeController.setVolume(volume, showSystemUI)
        result.success(null)
      }
      MethodName.IS_MUTED -> {
        val isMute = volumeController.isMute()
        result.success(isMute)
      }
      MethodName.SET_MUTE -> {
        val isMute: Boolean = call.argument(MethodArgument.IS_MUTE)!!
        val showSystemUI: Boolean = call.argument(MethodArgument.SHOW_SYSTEM_UI)!!

        volumeController.setMute(isMute, showSystemUI)
        result.success(null)
      }
      else -> result.notImplemented()
    }
  }

  override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    methodChannel.setMethodCallHandler(null)
    eventChannel.setStreamHandler(null)
  }
}

File: ios/volume_controller/Sources/volume_controller/VolumeControllerPlugin.swift
import AVFoundation
import Flutter
import UIKit

public class VolumeControllerPlugin: NSObject, FlutterPlugin {
  private let audioSession = AVAudioSession.sharedInstance()
  private lazy var volumeController = VolumeController(audioSession: audioSession)
  private lazy var volumeListener = VolumeListener(audioSession: audioSession)

  private func invalidArgumentsResult() -> FlutterError {
    FlutterError(
      code: "invalid_arguments",
      message: "Missing or invalid method arguments.",
      details: nil)
  }

  public static func register(with registrar: FlutterPluginRegistrar) {
    let methodChannel = FlutterMethodChannel(
      name: ChannelName.methodChannel, binaryMessenger: registrar.messenger())
    let instance = VolumeControllerPlugin()
    registrar.addMethodCallDelegate(instance, channel: methodChannel)

    let eventChannel = FlutterEventChannel(
      name: ChannelName.eventChannel,
      binaryMessenger: registrar.messenger())
    eventChannel.setStreamHandler(instance.volumeListener)

    registrar.addApplicationDelegate(instance)
    if #available(iOS 13.0, *) {
      registrar.addSceneDelegate(instance)
    }
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case MethodName.getVolume:
      result(volumeController.getVolume())
    case MethodName.setVolume:
      let arg = call.arguments as? [String: Any]
      let volume = arg?[MethodArgument.volume] as? Double
      let showSystemUI = arg?[MethodArgument.showSystemUI] as? Bool

      guard let volume, let showSystemUI else {
        result(invalidArgumentsResult())
        return
      }

      volumeController.setVolume(volume: Float(volume), showSystemUI: showSystemUI) { error in
        result(error)
      }
    case MethodName.isMuted:
      result(volumeController.isMuted())
    case MethodName.setMute:
      let arg = call.arguments as? [String: Any]
      let isMute = arg?[MethodArgument.isMute] as? Bool
      let showSystemUI = arg?[MethodArgument.showSystemUI] as? Bool

      guard let isMute, let showSystemUI else {
        result(invalidArgumentsResult())
        return
      }

      volumeController.setMute(isMute: isMute, showSystemUI: showSystemUI) { error in
        result(error)
      }
    default:
      result(FlutterMethodNotImplemented)
    }
  }
}

extension VolumeControllerPlugin: FlutterApplicationLifeCycleDelegate {
  public func applicationWillEnterForeground(_ application: UIApplication) {
    if #available(iOS 13.0, *) {
      return
    }

    volumeListener.resumeVolumeObservation()
  }
}

@available(iOS 13.0, *)
extension VolumeControllerPlugin: FlutterSceneLifeCycleDelegate {
  public func sceneWillEnterForeground(_ scene: UIScene) {
    volumeListener.resumeVolumeObservation()
  }
}
